src/D2L.Bmx/UpdateHandler.cs

```diff
@@ -0,0 +1,129 @@
+using System.Formats.Tar;
+using System.IO.Compression;
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+namespace D2L.Bmx;
+
+internal class UpdateHandler {
+
+	public async Task HandleAsync() {
+		using var httpClient = new HttpClient();
+		GithubRelease? releaseData = await GithubRelease.GetLatestReleaseDataAsync();
+		Version? latestVersion = releaseData?.GetReleaseVersion();
+		if( latestVersion is null ) {
+			throw new BmxException( "Failed to grab the latest version" );
+		}
+
+		var localVersion = Assembly.GetExecutingAssembly().GetName().Version;
+		if( latestVersion <= localVersion ) {
+			Console.WriteLine( $"Already own the latest version {latestVersion}" );
+			return;
+		}
+
+		string archiveName = GetOSFileName();
+		string? downloadUrl = releaseData?.Assets?.FirstOrDefault( a => a.Name == archiveName )?.BrowserDownloadUrl;
+		if( string.IsNullOrWhiteSpace( downloadUrl ) ) {
+			throw new BmxException( "Failed to get update download url" );
+		}
+
+		string? currentFilePath = Environment.ProcessPath;
+		if( string.IsNullOrEmpty( currentFilePath ) ) {
+			throw new BmxException( "BMX could not update" );
+		}
+
+		string currentDirectory = Path.GetDirectoryName( currentFilePath )!;
+		long backupPathTimeStamp = DateTime.Now.Millisecond;
+		string backupPath = Path.Join( BmxPaths.OLD_BMX_VERSIONS_PATH, $"bmx-v{localVersion}-{backupPathTimeStamp}-old.bak" );
+		try {
+			File.Move( currentFilePath, backupPath );
+		} catch( IOException ex ) {
+			throw new BmxException( "Could move the old version, try with elevated permissions", ex );
+		} catch {
+			throw new BmxException( "BMX could not update" );
+		}
+
+
+		string downloadPath = Path.GetTempFileName();
+		try {
+			var archiveRes = await httpClient.GetAsync( downloadUrl, HttpCompletionOption.ResponseHeadersRead );
+			using var fs = new FileStream( downloadPath, FileMode.Create, FileAccess.Write, FileShare.None );
+			await archiveRes.Content.CopyToAsync( fs );
+			await fs.FlushAsync();
+		} catch( Exception ex ) {
+			File.Move( backupPath, currentFilePath );
```

Comment:
Instead of moving the current file to backup path first then moving it back on error, how about downloading and extracting the new file into temp directory first, then do all the moves afterwards?

Reply:
I think for the TarFile extract we don't set or return the file name so I think it might be a bit complicated keeping track of which files to move in the directory if there are multiple updates running.

Reply:
Sorry, when I said "temp directory" I was actually thinking of a uniquely named subdirectory in the system temp directory, i.e. what's returned by [`Path.GetTempFileName`](https://learn.microsoft.com/en-us/dotnet/api/system.io.path.gettempfilename).

Even if we use `~/.bmx/temp`, we can also create more subdirectories in it with unique names, e.g.
```
~/.bmx/temp/58ba67a7-10f3-418c-a686-90894c2b6045
~/.bmx/temp/b90477fc-469b-4651-8ace-0ab522c935c8
...
```
we can generate and keep track of the UUID (or other types of unique identifiers e.g. timestamps?) in memory, and not worry about multiple runs interfering with each other.

Reply:
I made a commit trying out the temp directory first to push moving files to the end, seems a bit messy so I'm not sure if we should go for this. We can revert if needed

Reply:
The new change looks good. We can make it cleaner, but even as is it doesn't look too messy to me.
I do vastly prefer the new approach of "not moving files until we have everything ready".